lib/Page.js

```diff
@@ -632,6 +634,62 @@ class Page extends EventEmitter {
     return responses.get(this.mainFrame().url()) || null;
   }
 
+  /**
+   * @param {!Object=} options
+   * @return {!Promise<!Puppeteer.Request>}
+   */
+  async waitForRequest(url, options = {}) {
+    let matchedRequest;
+
+    const timeout = typeof options.timeout === 'number' ? options.timeout : this._defaultNetworkTimeout;
+    const watcher = new NetworkWatcher(this._networkManager, 'request', url, timeout, options);
```

Comment:
@matthewshirley I see the `NetworkWatcher` is modeled after `NavigationWatcher`. However, I believe this design is not needed in this case. Let me suggest an alternative approach.

Consider there's a function `waitForEvent` that returns a promise, such as:
- promise is fulfilled if `emitter` emits `event` that matches `predicate`
- promise is rejected if such an event is not emitted during `timeout` milliseconds

A draft implementation might look like this: 
```js
function waitForEvent(emitter, eventName, predicate, timeout) {
  let resolveCallback, rejectCallback;
  const promise = new Promise((resolve, reject) => {
    resolveCallback = resolve;
    rejectCallback = reject;
  });
  const listener = helper.addEventListener(emitter, eventName, event => {
    if (!predicate(event))
      return;
    cleanup();
    resolveCallback(event);
  });
  const timeoutId = setTimeout(() => {
    cleanup();
    rejectCallback(new Error('Timeout exceeded while waiting on request'));
  }, timeout);

  return promise;

  function cleanup() {
    helper.removeEventListeners([listener]);
    clearTimeout(timeoutId);
  }
}
```

Given this function, both `waitForRequest` and `waitForResponse` would be implementable like this:
```js
async waitForRequest(url, options = {}) {
  const timeout = typeof options.timeout === 'number' ? options.timeout : 30000;
  return waitForEvent(this._newtorkManager, NetworkManager.Events.Request, request => {
    return url === request.url() || (url instanceof RegExp && url.test(request.url()));
  }, timeout);
}

async waitForResponse(url, options = {}) {
  const timeout = typeof options.timeout === 'number' ? options.timeout : 30000;
  return waitForEvent(this._newtorkManager, NetworkManager.Events.Response, response => {
    return url === response.url() || (url instanceof RegExp && url.test(response.url()));
  }, timeout);
}
```

Reply:
Oh, awesome. Reviewing back my original method was a bit too much. That's a solid draft implementation and I couldn't find much to change. I've put this into the helper file.

I've reviewed the original issue raised and added the method option to both waitForRequest and waitForResponse and added statusCode to waitForResponse. Both can take an array.

Thanks!